### DISCUSSION TITLE
C# DTOs generate code in other languages with type errors

### DISCUSSION THREAD

**Discussion by mylvghb** [ServiceStack/Discuss#78] [General]
C# DTOs:
```CSharp

namespace WebApp.ServiceModel
{
    [Route("/contacts")]
    [DataContract]
    public class GetContacts
        : IReturn<GetContactsResponse>
    {
    }

    [DataContract]
    public class GetContactsResponse
        : Response<List<Contact>>
    {
    }

    [DataContract]
    public class Response<T>
    {
        [DataMember(Order=1)]
        public T Payload { get; set; }

        [DataMember(Order=2)]
        public ResponseStatus ResponseStatus { get; set; }
    }

}
```

Generated Typescript Code: 

```Typescript
// @DataContract
export class Response<T>
{
    // @DataMember(Order=1)
    public payload?: T;

    // @DataMember(Order=2)
    public responseStatus?: ResponseStatus;

    public constructor(init?: Partial<Response<T>>) { (Object as any).assign(this, init); }
}

// @DataContract
export class GetContactsResponse extends Array<Response<Contact>
{

    public constructor(init?: Partial<GetContactsResponse>) { super(); (Object as any).assign(this, init); }
}

// @Route("/contacts")
// @DataContract
export class GetContacts implements IReturn<GetContactsResponse>
{

    public constructor(init?: Partial<GetContacts>) { (Object as any).assign(this, init); }
    public getTypeName() { return 'GetContacts'; }
    public getMethod() { return 'GET'; }
    public createResponse() { return new GetContactsResponse(); }
}
```

The response DTO `GetContactsResponse` should be extended from `Response<Array<Contact>>` or `Response<Contact[]>`, not `Array<Response<Contact>>`.

Additionally, I tried to change List<> to IList<> or IEnumerable<>, but the automatically generated code type is also IList<> or IEnumerable<>. However, some languages do not have this type.

**Comment by mylvghb** [comment]
Js also has the same issue:

```javascript
export class GetContactsResponse extends Array {
    /** @param {{payload?:T,responseStatus?:ResponseStatus}} [init] */
    constructor(init) { super(init); Object.assign(this, init) }
}
```

**Comment by mythz** [comment]
We don't support inheriting from nested generic types, change the DTO so the `List<T>` is in the class definition, i.e:

```csharp
[DataContract]
public class Response<T>
{
    [DataMember(Order=1)]
    public List<T> Payload { get; set; }

    [DataMember(Order=2)]
    public ResponseStatus ResponseStatus { get; set; }
}

[DataContract]
public class GetContactsResponse
    : Response<Contact>
{
}
```

**Reply by mylvghb** [reply]
Dart, Java And Kotlion etc..., generated is right.
![image](https://github.com/ServiceStack/Discuss/assets/35945793/c6ee0d10-9749-4458-a9c8-718455453d1e)
![image](https://github.com/ServiceStack/Discuss/assets/35945793/b00a5829-365e-4a3c-b8d1-28d14aaf90e8)
![image](https://github.com/ServiceStack/Discuss/assets/35945793/00f3077d-bc8b-4230-9368-1d2ae6b2654a)

**Reply by mylvghb** [reply]
I tried your code and the generated TS/JS types were correct, but if I change the type to IList<> or IEnumerable<> in the Response, some languages will have type errors also.

**Reply by mythz** [reply]
You should [never use interface properties in DTOs](https://stackoverflow.com/a/10759250/85785), always use a concrete collection like `List<T>`.

**Reply by mylvghb** [reply]
Thx, I understand.
